src/test/java/com/github/thorbenkuck/netcom2/network/shared/SessionImplTest.java

```diff
@@ -1,66 +1,356 @@
+		assertNotNull(pipeline);
+		assertEquals(pipeline, session.eventOf(TestObject.class));
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void eventOfNull() throws Exception {
+		// Arrange
+		SessionImpl session = new SessionImpl(bridge);
+
+		// Act
+		session.eventOf(null);
+
+		// Assert
+		fail();
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void triggerEventNotRegistered() throws Exception {
+		// Arrange
+		SessionImpl session = new SessionImpl(bridge);
+
+		// Act
+		session.triggerEvent(TestObject.class, new TestObject());
+
+		// Assert
+		fail();
 	}
 
 	@Test
 	public void triggerEvent() throws Exception {
+		// Arrange
+		SessionImpl session = new SessionImpl(bridge);
+		TestObject testObject = new TestObject();
+		AtomicBoolean finished = new AtomicBoolean(false);
+		session.eventOf(TestObject.class).addFirst(o -> {
+			assertEquals(testObject, o);
+			finished.set(true);
+		});
+
+		// Act
+		session.triggerEvent(TestObject.class, testObject);
+
+		// Assert
+		assertTrue(finished.get());
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void triggerEventObjectNull() throws Exception {
+		// Arrange
+		SessionImpl session = new SessionImpl(bridge);
+
+		// Act
+		session.triggerEvent(TestObject.class, null);
+
+		// Assert
+		fail();
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void triggerEventClassNull() throws Exception {
+		// Arrange
+		SessionImpl session = new SessionImpl(bridge);
+
+		// Act
+		session.triggerEvent(null, new TestObject());
+
+		// Assert
+		fail();
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void triggerEventClassAndObjectNull() throws Exception {
+		// Arrange
+		SessionImpl session = new SessionImpl(bridge);
+
+		// Act
+		session.triggerEvent(null, new TestObject());
+
+		// Assert
+		fail();
 	}
 
 	@Test
 	public void addHeartBeat() throws Exception {
+		// Arrange
+		SessionImpl session = new SessionImpl(bridge);
+		HeartBeat<Session> heartBeat = mock(HeartBeat.class);
+		HeartBeatParallel<Session> parallel = mock(HeartBeatParallel.class);
+		when(heartBeat.parallel()).thenReturn(parallel);
+
+		// Act
+		session.addHeartBeat(heartBeat);
+
+		// Assert
+		verify(heartBeat).parallel();
+		verify(parallel).run(eq(session));
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void addHeartBeatNull() throws Exception {
+		// Arrange
+		SessionImpl session = new SessionImpl(bridge);
+
+		// Act
+		session.addHeartBeat(null);
+
+		// Assert
+		fail();
 	}
 
 	@Test
-	public void removeHeartBeat() throws Exception {
+	public void removeHeartBeatNotRegistered() throws Exception {
+		// Arrange
+		Logging logging = mock(Logging.class);
+		NetComLogging.setLogging(logging);
+		SessionImpl session = new SessionImpl(bridge);
+		HeartBeat<Session> heartBeat = mock(HeartBeat.class);
+
+		// Act
+		session.removeHeartBeat(heartBeat);
+
+		// Assert
+		verify(logging).warn(contains("was never set"));
```

Comment:
Is this really a good idea?

Reply:
I added a never check in Heartbeat#stop. but the reason, this logging check is in here, is that removeHeartBeat returns directly if the heartBeat is not set. The only thing this does is, that it loggs a warning to the developer. Since the Logging-API is interchangable, this means, the developer should receive any notification about that the HeartBeat was never set.

In the other parts of NetCom2, this would be done, using an Exception. But this is not applicable here, since this is not an Illegal State, an Illegal Argument or something else